include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+           NonlinearSolverSelector<VectorType>::AdditionalData::newton)
+    additional_data_kinsol.strategy =
+      SUNDIALS::KINSOL<VectorType>::AdditionalData::newton;
+  else if (additional_data.strategy ==
+           NonlinearSolverSelector<VectorType>::AdditionalData::picard)
+    additional_data_kinsol.strategy =
+      SUNDIALS::KINSOL<VectorType>::AdditionalData::picard;
+
+  // Setting data points in the KINSOL class from the NonlinearSolverSelector
+  // class
+  additional_data_kinsol.maximum_non_linear_iterations =
+    additional_data.maximum_non_linear_iterations;
+  additional_data_kinsol.function_tolerance =
+    additional_data.function_tolerance;
+  additional_data_kinsol.step_tolerance = additional_data.step_tolerance;
+  additional_data_kinsol.anderson_subspace_size =
+    additional_data.anderson_subspace_size;
+#endif
+
+// Do the same thing we did above but with NOX
+#ifdef DEAL_II_WITH_TRILINOS
+  parameters_nox->set("Nonlinear Solver", "Line Search Based");
+  Teuchos::ParameterList &Line_Search = parameters_nox->sublist("Line Search");
+  Line_Search.set("Method", "Full Step");
+
+  additional_data_nox.max_iter = additional_data.maximum_non_linear_iterations;
+  additional_data_nox.abs_tol  = additional_data.function_tolerance;
+  additional_data_nox.rel_tol  = additional_data.relative_tolerance;
+#endif
+
+#ifdef DEAL_II_WITH_PETSC
+  additional_data_petsc_snes.options_prefix = "";
+
+  if (additional_data.anderson_subspace_size > 0 &&
+      additional_data.strategy ==
+        NonlinearSolverSelector<VectorType>::AdditionalData::picard)
+    {
+      additional_data_petsc_snes.snes_type = "anderson";
+      // TODO additional_data.anderson_subspace_size;
+    }
+  else if (additional_data.strategy ==
+           NonlinearSolverSelector<VectorType>::AdditionalData::linesearch)
+    {
+      additional_data_petsc_snes.snes_type            = "newtonls";
+      additional_data_petsc_snes.snes_linesearch_type = "bt";
+    }
+  else if (additional_data.strategy ==
+           NonlinearSolverSelector<VectorType>::AdditionalData::newton)
+    {
+      additional_data_petsc_snes.snes_linesearch_type = "newtonls";
+      additional_data_petsc_snes.snes_linesearch_type = "basic";
+    }
+  else if (additional_data.strategy ==
+           NonlinearSolverSelector<VectorType>::AdditionalData::picard)
+    additional_data_petsc_snes.snes_type = "nrichardson";
+
+  additional_data_petsc_snes.absolute_tolerance =
+    additional_data.function_tolerance;
+  additional_data_petsc_snes.relative_tolerance =
+    additional_data.relative_tolerance;
+  additional_data_petsc_snes.step_tolerance = additional_data.step_tolerance;
+  additional_data_petsc_snes.maximum_non_linear_iterations =
+    additional_data.maximum_non_linear_iterations;
+  additional_data_petsc_snes.max_n_function_evaluations = -1;
+
+#endif
+}
+
+template <typename VectorType>
+NonlinearSolverSelector<VectorType>::NonlinearSolverSelector() = default;
```

Comment:
Doesn't this leave a whole set of `additional_data_*` objects uninitialized?

Reply:
All the ``AdditionalData`` classes have default constructors that set defaults, as far as I know.